deerlab/dipolarkernel.py

```diff
@@ -213,76 +233,104 @@ def dipolarkernel(t, r, *, pathways=None, mod=None, bg=None, method='fresnel', e
+    # Vectorized 3D-grid evaluation (t,r,powder averaging)
+    C = np.cos(ωr[np.newaxis,:,np.newaxis]*q[np.newaxis,np.newaxis,:]*abs(t[:,np.newaxis,np.newaxis]))
+    # If given, include limited excitation bandwidth
+    if not np.isinf(ωex):
+        C = C*np.exp(-(ωr[np.newaxis,:,np.newaxis]*q[np.newaxis,np.newaxis,:])**2/ωex**2)
+    return C
 #==============================================================================
 
+@cached(max_size=100)
+def elementarykernel(t,r,method,ωex,nKnots,g,Pθ):
+#==============================================================================
+    "Calculates the elementary dipolar kernel (cached for speed)"
+
     t = np.atleast_1d(t)
     r = np.atleast_1d(r)
 
     # Pre-allocate K matrix
     nr = np.size(r)
     nt = np.size(t)  
-    K = np.zeros((nt,nr))
-    wr = w0(g)/(r**3)  # rad s^-1
+    K0 = np.zeros((nt,nr))
+    ωr = ω0(g)/(r**3)  # rad μs^-1
+    
+    orientationselection = Pθ is not None 
 
-    def kernelmatrix_fresnel(t,r):
+
+    def elementarykernel_fresnel(t):
     #==========================================================================
         """Calculate kernel using Fresnel integrals (fast and accurate)"""
 
        # Calculation using Fresnel integrals
-        ph = np.outer(np.abs(t), wr)
-        kappa = np.sqrt(6*ph/pi)
+        ɸ = np.outer(np.abs(t), ωr)
+        κ = np.sqrt(6*ɸ/π)
         
         # Supress divide by 0 warning        
         with warnings.catch_warnings():
             warnings.simplefilter('ignore')
-            S, C = fresnel(kappa)/kappa
+            S, C = fresnel(κ)/κ
         
-        K = C*np.cos(ph) + S*np.sin(ph)
-        K[t==0] = 1 
+        K0 = C*np.cos(ɸ) + S*np.sin(ɸ)
+
+        # Limit of K0(t,r) as t->0
+        K0[t==0] = 1 
 
-        return K
+        return K0
     #==========================================================================
-    
-    def kernelmatrix_grid(t,r,wex,nKnots):
+
+    def elementarykernel_grid(t,ωex,nKnots,Pθ):
     #==========================================================================
         """Calculate kernel using grid-based powder integration (converges very slowly with nKnots)"""
 
         # Orientational grid
-        costheta = np.linspace(0,1,int(nKnots))
-        q = 1 - 3*costheta**2
-        # Vectorized 3D-grid evaluation (t,r,powder averaging)
-        C = np.cos(wr[np.newaxis,:,np.newaxis]*q[np.newaxis,np.newaxis,:]*abs(t[:,np.newaxis,np.newaxis]))
-        # If given, include limited excitation bandwidth
-        if not np.isinf(wex):
-            C = C*np.exp(-(wr[np.newaxis,:,np.newaxis]*q[np.newaxis,np.newaxis,:])**2/wex**2)
-        K = np.sum(C,2)/nKnots
-        return K
+        z = np.linspace(0,1,int(nKnots))
+        θ = np.arccos(z) # rad
+        q = 1 - 3*z**2
+
+        if orientationselection:
+            # Evaluate orientational distribution over grid        
+            Pθnorm,_ = scipy.integrate.quad(lambda z: Pθ(np.arccos(z)),0,1,limit=1000)
```

Comment:
Based on symmetry, this interpolant must have zero derivative at 0 and pi/2

Reply:
It was not strictly satisfied. It depended on whether the user enforced that on the input function. Now `dipolarkernel` constructs a constrained spline from the input distribution to ensure zero derivatives at 0 and pi/2.